File: 05_practice/src/com/gomboclabs/codespirit/practice/basic/class01/go/Code02_BubbleSort.go
package main

// 你可以将其包名设置为main。这样，它将成为一个独立的可执行文件，
//并且可以通过go build命令构建和运行。

import (
	"fmt"
	"math/rand"
	"sort"
	"time"
)

func bubbleSort(arr []int) {
	size := len(arr)
	if size < 2 {
		return
	}

	for i := 0; i < size; i++ {
		isSwapped := false
		for j := size - 1; j > i; j-- {
			if arr[j] < arr[j-1] {
				isSwapped = true
				arr[j], arr[j-1] = arr[j-1], arr[j]
			}
		}
		if isSwapped != true {
			break
		}
	}
}

func comparator(arr []int) {
	sort.Ints(arr)
}

func generateRandomArray(size, maxValue int) []int {
	arr := make([]int, size)
	rand.Seed(time.Now().UnixNano())
	for i := 0; i < size; i++ {
		arr[i] = rand.Intn(2*maxValue+1) - maxValue
	}
	return arr
}

func copyArray(src []int) []int {
	dest := make([]int, len(src))
	copy(dest, src)
	return dest
}

func isEqual(arr1, arr2 []int) bool {
	if len(arr1) != len(arr2) {
		return false
	}

	for i := 0; i < len(arr1); i++ {
		if arr1[i] != arr2[i] {
			return false
		}
	}
	return true
}

func printArray(arr []int) {
	for _, num := range arr {
		fmt.Printf("%d ", num)
	}
	fmt.Println()
}

func main() {
	testTime := 500000
	maxSize := 100
	maxValue := 100
	succeed := true

	// Start the timer
	start := time.Now()

	for i := 0; i < testTime; i++ {
		size := rand.Intn(maxSize) + 1
		arr1 := generateRandomArray(size, maxValue)
		arr2 := copyArray(arr1)

		bubbleSort(arr1)
		comparator(arr2)

		if !isEqual(arr1, arr2) {
			succeed = false
			printArray(arr1)
			printArray(arr2)
			break
		}
	}

	if succeed {
		fmt.Println("Nice!")
	} else {
		fmt.Println("Fucking fucked!")
	}

	// Stop the timer
	elapsed := time.Since(start)

	// Print the execution time
	fmt.Printf("Execution time: %s\n", elapsed)
}

File: 05_practice/src/com/gomboclabs/codespirit/practice/basic/class01/scala/Code02_BubbleSort.scala
package gomboclabs.codespirit.practice.basic.class01.scala

import scala.util.Random

object Code02_BubbleSort {

  //  练手
  def bubbleSort(arr: Array[Int]): Unit = {
    var swapped = false;
    for (i <- 0 until arr.length  by 1) {
      for (j <- arr.length - 1 until i by -1) {
        if (arr(j) < arr(j - 1)){
          swapped = true
          swap(arr, j, j-1)
        }
      }
      // for (i <- 0 until arr.length - 1 by 1 if !isSorted) {} 可使用if替代
      if (!swapped) {
        return
      }
    }
  }

  def swap(arr: Array[Int], i: Int, j: Int): Unit = {
    val tmp = arr(i)
    arr(i) = arr(j)
    arr(j) = tmp
  }

  // for test
  def comparator(arr: Array[Int]): Unit = {
    arr.sorted
  }

  // for test
  def generateRandomArray(maxSize: Int, maxValue: Int): Array[Int] = {
    val random = new Random()
    val size = random.nextInt(maxSize + 1)
    Array.fill(size)(random.nextInt(2 * maxValue + 1) - maxValue)
  }

  // for test
  def copyArray(arr: Array[Int]): Array[Int] = {
    if (arr == null) {
      null
    } else {
      arr.clone()
    }
  }

  // for test
  def isEqual(arr1: Array[Int], arr2: Array[Int]): Boolean = {
    if (arr1 == null && arr2 == null) {
      true
    } else if (arr1 == null || arr2 == null) {
      false
    } else {
      arr1.sameElements(arr2)
    }
  }

  // for test
  def printArray(arr: Array[Int]): Unit = {
    if (arr != null) {
      println(arr.mkString(" "))
    }
  }

  // for test
  def main(args: Array[String]): Unit = {
    val startTime = System.nanoTime()
    val testTime = 500000
    val maxSize = 100
    val maxValue = 100
    val succeed = true

    for (_ <- 0 until testTime) {
      val arr1 = generateRandomArray(maxSize, maxValue)
      val arr2 = copyArray(arr1)

      bubbleSort(arr1)
      comparator(arr2)
    }

    println(if (succeed) "Nice!" else "Fucking fucked!")
    val endTime = System.nanoTime()
    val executionTime = (endTime - startTime) / 1000000

    println(s"执行时间：$executionTime 毫秒")
  }
}